### PR TITLE
[ASM] Added 1 hour timer to vulnerability deduplication

### PR DESCRIPTION
**Pull request by NachoEchevarria**
## Summary of changes

Vulnerability deduplication stores the detected vulnerabilities and reports them only once per execution. A 1 hour timer is required to be set in order to clear the deduplication cache periodically. That will mean that every hour, repeated vulnerabilities might be sent. A single timer will be used for the whole cache. The individual vulnerabilities have no timer associated.

### CODE REVIEW

**Comment by andrewlock** [comment]
## Execution-Time Benchmarks Report :stopwatch:

Execution-time results for samples comparing the following branches/commits:
- [This PR (4411)](https://github.com/DataDog/dd-trace-dotnet/tree/88300bb4ee41c70fbf931a8bb5632120c7f29efd)
- [master](https://github.com/DataDog/dd-trace-dotnet/tree/4bafc78fa25382fefb03fa4e052905ca371d1ffa)

Execution-time benchmarks measure the whole time it takes to execute a program. And are intended to measure the one-off costs. Cases where the execution time results for the PR are worse than latest master results are shown in **red**. The following thresholds were used for comparing the execution times:
* Welch test with statistical test for significance of **5%**
* Only results indicating a difference greater than **5%** and **5 ms** are considered.


Note that these results are based on a _single_ point-in-time result for each branch. For full results, see the [dashboard](https://ddstaging.datadoghq.com/dashboard/4qn-6fi-54p/apm-net-execution-time-benchmarks).

Graphs show the p99 interval based on the mean and StdDev of the test run, as well as the mean value of the run (shown as a diamond below the graph).
```mermaid
gantt
    title Execution time (ms) FakeDbCommand (.NET Framework 4.6.2) 
    dateFormat  X
    axisFormat %s
    todayMarker off
    section Baseline
    This PR (4411) - mean (3,006ms)  : 2946, 3066
     .   : milestone, 3006,
    master - mean (2,985ms)  : 2892, 3078
     .   : milestone, 2985,

    section CallTarget+Inlining+NGEN
    This PR (4411) - mean (3,822ms)  : 3731, 3914
     .   : milestone, 3822,
    master - mean (3,813ms)  : 3735, 3891
     .   : milestone, 3813,

```
```mermaid
gantt
    title Execution time (ms) FakeDbCommand (.NET Core 3.1) 
    dateFormat  X
    axisFormat %s
    todayMarker off
    section Baseline
    This PR (4411) - mean (3,123ms)  : 3017, 3230
     .   : milestone, 3123,
    master - mean (3,104ms)  : 3020, 3188
     .   : milestone, 3104,

    section CallTarget+Inlining+NGEN
    This PR (4411) - mean (3,603ms)  : 3553, 3653
     .   : milestone, 3603,
    master - mean (3,617ms)  : 3547, 3687
     .   : milestone, 3617,

```
```mermaid
gantt
    title Execution time (ms) FakeDbCommand (.NET 6) 
    dateFormat  X
    axisFormat %s
    todayMarker off
    section Baseline
    This PR (4411) - mean (3,087ms)  : 2997, 3176
     .   : milestone, 3087,
    master - mean (3,109ms)  : 2988, 3230
     .   : milestone, 3109,

    section CallTarget+Inlining+NGEN
    This PR (4411) - mean (3,591ms)  : 3516, 3666
     .   : milestone, 3591,
    master - mean (3,562ms)  : 3502, 3621
     .   : milestone, 3562,

```
```mermaid
gantt
    title Execution time (ms) HttpMessageHandler (.NET Framework 4.6.2) 
    dateFormat  X
    axisFormat %s
    todayMarker off
    section Baseline
    This PR (4411) - mean (191ms)  : 187, 195
     .   : milestone, 191,
    master - mean (191ms)  : 187, 195
     .   : milestone, 191,

    section CallTarget+Inlining+NGEN
    This PR (4411) - mean (1,115ms)  : 1084, 1147
     .   : milestone, 1115,
    master - mean (1,126ms)  : 1090, 1163
     .   : milestone, 1126,

```
```mermaid
gantt
    title Execution time (ms) HttpMessageHandler (.NET Core 3.1) 
    dateFormat  X
    axisFormat %s
    todayMarker off
    section Baseline
    This PR (4411) - mean (373ms)  : 364, 382
     .   : milestone, 373,
    master - mean (372ms)  : 367, 377
     .   : milestone, 372,

    section CallTarget+Inlining+NGEN
    This PR (4411) - mean (1,169ms)  : 1144, 1194
     .   : milestone, 1169,
    master - mean (1,177ms)  : 1148, 1205
     .   : milestone, 1177,

```
```mermaid
gantt
    title Execution time (ms) HttpMessageHandler (.NET 6) 
    dateFormat  X
    axisFormat %s
    todayMarker off
    section Baseline
    This PR (4411) - mean (359ms)  : 353, 365
     .   : milestone, 359,
    master - mean (361ms)  : 357, 366
     .   : milestone, 361,

    section CallTarget+Inlining+NGEN
    This PR (4411) - mean (1,131ms)  : 1105, 1156
     .   : milestone, 1131,
    master - mean (1,129ms)  : 1105, 1152
     .   : milestone, 1129,

```

**Comment by andrewlock** [comment]
## Throughput/Crank Report:zap:

Throughput results for AspNetCoreSimpleController comparing the following branches/commits:
- [This PR (4411)](https://github.com/DataDog/dd-trace-dotnet/tree/88300bb4ee41c70fbf931a8bb5632120c7f29efd)
- [master](https://github.com/DataDog/dd-trace-dotnet/tree/4bafc78fa25382fefb03fa4e052905ca371d1ffa)
- [benchmarks/2.9.0](https://github.com/DataDog/dd-trace-dotnet/tree/89ae3f4ee7d85b824badda5873166707c9e560b4)
- [benchmarks/2.33.0](https://github.com/DataDog/dd-trace-dotnet/tree/54061c090c4447852d373c5c7f39b9dcc7746f19)

Cases where throughput results for the PR are worse than latest master (5% drop or greater), results are shown in **red**.

Note that these results are based on a _single_ point-in-time result for each branch. For full results, see [one](https://ddstaging.datadoghq.com/dashboard/fnz-afi-c2i/apm-net-crank) of the [many](https://ddstaging.datadoghq.com/dashboard/c5g-i7d-kcy/ciapp-apm-net-throughput-tests), [many](https://ddstaging.datadoghq.com/dashboard/uxh-m8j-qhi/ciapp-apm-net-throughput-tests-kevin) dashboards!

```mermaid
gantt
    title Throughput Linux x64 (Total requests) 
    dateFormat  X
    axisFormat %s
    section Baseline
    This PR (4411) (11.553M)   : 0, 11552534
    master (11.160M)   : 0, 11160008
    benchmarks/2.33.0 (10.950M)   : 0, 10949608
    benchmarks/2.9.0 (11.142M)   : 0, 11141669

    section Automatic
    This PR (4411) (8.082M)   : 0, 8081894
    master (7.870M)   : 0, 7870164
    benchmarks/2.33.0 (7.621M)   : 0, 7620707
    benchmarks/2.9.0 (7.959M)   : 0, 7958947

    section Trace stats
    master (7.862M)   : 0, 7862341
    benchmarks/2.33.0 (7.602M)   : 0, 7602038

    section Manual
    This PR (4411) (10.335M)   : 0, 10335142
    master (10.104M)   : 0, 10104483
    benchmarks/2.33.0 (9.770M)   : 0, 9770235

    section Manual + Automatic
    This PR (4411) (7.778M)   : 0, 7778339
    master (7.514M)   : 0, 7514200
    benchmarks/2.33.0 (7.189M)   : 0, 7188658

    section Version Conflict
    master (6.684M)   : 0, 6683701
    benchmarks/2.33.0 (6.553M)   : 0, 6553316

```
```mermaid
gantt
    title Throughput Linux arm64 (Total requests) 
    dateFormat  X
    axisFormat %s
    section Baseline
    This PR (4411) (9.704M)   : 0, 9704433
    master (9.834M)   : 0, 9834025
    benchmarks/2.33.0 (9.426M)   : 0, 9425603
    benchmarks/2.9.0 (9.842M)   : 0, 9842216

    section Automatic
    This PR (4411) (6.683M)   : 0, 6683401
    master (6.734M)   : 0, 6734121
    benchmarks/2.33.0 (6.803M)   : 0, 6802932

    section Trace stats
    master (6.663M)   : 0, 6662679
    benchmarks/2.33.0 (6.810M)   : 0, 6810447

    section Manual
    This PR (4411) (8.359M)   : 0, 8359441
    master (8.556M)   : 0, 8555955
    benchmarks/2.33.0 (8.435M)   : 0, 8434569

    section Manual + Automatic
    This PR (4411) (6.580M)   : 0, 6579825
    master (6.512M)   : 0, 6511692
    benchmarks/2.33.0 (6.465M)   : 0, 6465149

    section Version Conflict
    master (5.908M)   : 0, 5907795
    benchmarks/2.33.0 (5.807M)   : 0, 5806690

```
```mermaid
gantt
    title Throughput Windows x64 (Total requests) 
    dateFormat  X
    axisFormat %s
    section Baseline
    This PR (4411) (10.724M)   : 0, 10724285
    master (10.573M)   : 0, 10573116
    benchmarks/2.33.0 (9.426M)   : 0, 9426080
    benchmarks/2.9.0 (10.688M)   : 0, 10687842

    section Automatic
    This PR (4411) (7.349M)   : 0, 7349393
    master (7.688M)   : 0, 7687501
    benchmarks/2.33.0 (6.469M)   : 0, 6468939
    benchmarks/2.9.0 (7.829M)   : 0, 7828890

    section Trace stats
    master (7.554M)   : 0, 7553892
    benchmarks/2.33.0 (6.830M)   : 0, 6830314

    section Manual
    This PR (4411) (9.396M)   : 0, 9395687
    master (9.655M)   : 0, 9654627
    benchmarks/2.33.0 (8.108M)   : 0, 8108026

    section Manual + Automatic
    This PR (4411) (7.481M)   : 0, 7481356
    master (7.450M)   : 0, 7449965
    benchmarks/2.33.0 (6.856M)   : 0, 6856379

    section Version Conflict
    master (6.667M)   : 0, 6666913
    benchmarks/2.33.0 (5.893M)   : 0, 5893016

```
```mermaid
gantt
    title Throughput Linux x64 (ASM) (Total requests) 
    dateFormat  X
    axisFormat %s
    section Baseline
    This PR (4411) (7.618M)   : 0, 7617982
    master (7.320M)   : 0, 7320437
    benchmarks/2.33.0 (7.765M)   : 0, 7764607
    benchmarks/2.9.0 (7.769M)   : 0, 7768524

    section No attack
    This PR (4411) (2.148M)   : 0, 2147923
    master (2.138M)   : 0, 2138218
    benchmarks/2.33.0 (2.209M)   : 0, 2208725
    benchmarks/2.9.0 (3.174M)   : 0, 3174364

    section Attack
    This PR (4411) (1.812M)   : 0, 1811945
    master (1.817M)   : 0, 1817182
    benchmarks/2.33.0 (1.911M)   : 0, 1910909
    benchmarks/2.9.0 (2.518M)   : 0, 2517764

    section Blocking
    This PR (4411) (3.710M)   : 0, 3710270
    master (3.694M)   : 0, 3694196
    benchmarks/2.33.0 (3.861M)   : 0, 3860900

```

File: tracer/src/Datadog.Trace/Iast/HashBasedDeduplication.cs [hunk 1]
```diff
@@ -3,38 +3,51 @@
 // This product includes software developed at Datadog (https://www.datadoghq.com/). Copyright 2017 Datadog, Inc.
 // </copyright>
 
+using System;
 using System.Collections.Generic;
 
 namespace Datadog.Trace.Iast;
```
Comment (daniel-romano-DD) [inline tracer/src/Datadog.Trace/Iast/HashBasedDeduplication.cs:9]: #nullable enable missing here
```suggestion
#nullable enable 
namespace Datadog.Trace.Iast;
```

**Review by daniel-romano-DD — APPROVED** [review]
LGTM with a NIT

**Review by robertpi — APPROVED** [review]

**Comment by andrewlock** [comment]
## Execution-Time Benchmarks Report :stopwatch:

Execution-time results for samples comparing the following branches/commits:
- [This PR (4411)](https://github.com/DataDog/dd-trace-dotnet/tree/66b9961e409221cd45d780662a3e73ef36018ffe)
- [master](https://github.com/DataDog/dd-trace-dotnet/tree/b44cd82e291f259df8ee58c7d117ecf32f6215de)

Execution-time benchmarks measure the whole time it takes to execute a program. And are intended to measure the one-off costs. Cases where the execution time results for the PR are worse than latest master results are shown in **red**. The following thresholds were used for comparing the execution times:
* Welch test with statistical test for significance of **5%**
* Only results indicating a difference greater than **5%** and **5 ms** are considered.


Note that these results are based on a _single_ point-in-time result for each branch. For full results, see the [dashboard](https://ddstaging.datadoghq.com/dashboard/4qn-6fi-54p/apm-net-execution-time-benchmarks).

Graphs show the p99 interval based on the mean and StdDev of the test run, as well as the mean value of the run (shown as a diamond below the graph).
```mermaid
gantt
    title Execution time (ms) FakeDbCommand (.NET Framework 4.6.2) 
    dateFormat  X
    axisFormat %s
    todayMarker off
    section Baseline
    This PR (4411) - mean (2,986ms)  : 2888, 3084
     .   : milestone, 2986,
    master - mean (2,998ms)  : 2926, 3070
     .   : milestone, 2998,

    section CallTarget+Inlining+NGEN
    This PR (4411) - mean (3,831ms)  : 3731, 3930
     .   : milestone, 3831,
    master - mean (3,817ms)  : 3735, 3899
     .   : milestone, 3817,

```
```mermaid
gantt
    title Execution time (ms) FakeDbCommand (.NET Core 3.1) 
    dateFormat  X
    axisFormat %s
    todayMarker off
    section Baseline
    This PR (4411) - mean (3,113ms)  : 3044, 3181
     .   : milestone, 3113,
    master - mean (3,118ms)  : 3028, 3207
     .   : milestone, 3118,

    section CallTarget+Inlining+NGEN
    This PR (4411) - mean (3,617ms)  : 3552, 3681
     .   : milestone, 3617,
    master - mean (3,614ms)  : 3548, 3680
     .   : milestone, 3614,

```
```mermaid
gantt
    title Execution time (ms) FakeDbCommand (.NET 6) 
    dateFormat  X
    axisFormat %s
    todayMarker off
    section Baseline
    This PR (4411) - mean (3,098ms)  : 3029, 3168
     .   : milestone, 3098,
    master - mean (3,113ms)  : 3027, 3199
     .   : milestone, 3113,

    section CallTarget+Inlining+NGEN
    This PR (4411) - mean (3,576ms)  : 3510, 3642
     .   : milestone, 3576,
    master - mean (3,571ms)  : 3515, 3627
     .   : milestone, 3571,

```
```mermaid
gantt
    title Execution time (ms) HttpMessageHandler (.NET Framework 4.6.2) 
    dateFormat  X
    axisFormat %s
    todayMarker off
    section Baseline
    This PR (4411) - mean (192ms)  : 186, 197
     .   : milestone, 192,
    master - mean (191ms)  : 187, 195
     .   : milestone, 191,

    section CallTarget+Inlining+NGEN
    This PR (4411) - mean (1,129ms)  : 1096, 1162
     .   : milestone, 1129,
    master - mean (1,123ms)  : 1089, 1157
     .   : milestone, 1123,

```
```mermaid
gantt
    title Execution time (ms) HttpMessageHandler (.NET Core 3.1) 
    dateFormat  X
    axisFormat %s
    todayMarker off
    section Baseline
    This PR (4411) - mean (375ms)  : 369, 382
     .   : milestone, 375,
    master - mean (375ms)  : 368, 381
     .   : milestone, 375,

    section CallTarget+Inlining+NGEN
    This PR (4411) - mean (1,179ms)  : 1158, 1199
     .   : milestone, 1179,
    master - mean (1,185ms)  : 1159, 1211
     .   : milestone, 1185,

```
```mermaid
gantt
    title Execution time (ms) HttpMessageHandler (.NET 6) 
    dateFormat  X
    axisFormat %s
    todayMarker off
    section Baseline
    This PR (4411) - mean (361ms)  : 356, 367
     .   : milestone, 361,
    master - mean (365ms)  : 357, 372
     .   : milestone, 365,

    section CallTarget+Inlining+NGEN
    This PR (4411) - mean (1,139ms)  : 1111, 1166
     .   : milestone, 1139,
    master - mean (1,146ms)  : 1117, 1175
     .   : milestone, 1146,

```

**Comment by andrewlock** [comment]
## Throughput/Crank Report:zap:

Throughput results for AspNetCoreSimpleController comparing the following branches/commits:
- [This PR (4411)](https://github.com/DataDog/dd-trace-dotnet/tree/66b9961e409221cd45d780662a3e73ef36018ffe)
- [master](https://github.com/DataDog/dd-trace-dotnet/tree/b44cd82e291f259df8ee58c7d117ecf32f6215de)
- [benchmarks/2.9.0](https://github.com/DataDog/dd-trace-dotnet/tree/89ae3f4ee7d85b824badda5873166707c9e560b4)
- [benchmarks/2.33.0](https://github.com/DataDog/dd-trace-dotnet/tree/54061c090c4447852d373c5c7f39b9dcc7746f19)

Cases where throughput results for the PR are worse than latest master (5% drop or greater), results are shown in **red**.

Note that these results are based on a _single_ point-in-time result for each branch. For full results, see [one](https://ddstaging.datadoghq.com/dashboard/fnz-afi-c2i/apm-net-crank) of the [many](https://ddstaging.datadoghq.com/dashboard/c5g-i7d-kcy/ciapp-apm-net-throughput-tests), [many](https://ddstaging.datadoghq.com/dashboard/uxh-m8j-qhi/ciapp-apm-net-throughput-tests-kevin) dashboards!

```mermaid
gantt
    title Throughput Linux x64 (Total requests) 
    dateFormat  X
    axisFormat %s
    section Baseline
    This PR (4411) (11.410M)   : 0, 11410333
    master (11.146M)   : 0, 11146357
    benchmarks/2.33.0 (10.950M)   : 0, 10949608
    benchmarks/2.9.0 (11.142M)   : 0, 11141669

    section Automatic
    This PR (4411) (7.897M)   : 0, 7897414
    master (7.839M)   : 0, 7839183
    benchmarks/2.33.0 (7.621M)   : 0, 7620707
    benchmarks/2.9.0 (7.959M)   : 0, 7958947

    section Trace stats
    master (7.839M)   : 0, 7838681
    benchmarks/2.33.0 (7.602M)   : 0, 7602038

    section Manual
    This PR (4411) (10.072M)   : 0, 10072121
    master (10.041M)   : 0, 10040618
    benchmarks/2.33.0 (9.770M)   : 0, 9770235

    section Manual + Automatic
    This PR (4411) (7.529M)   : 0, 7528889
    master (7.524M)   : 0, 7524092
    benchmarks/2.33.0 (7.189M)   : 0, 7188658

    section Version Conflict
    master (6.689M)   : 0, 6689193
    benchmarks/2.33.0 (6.553M)   : 0, 6553316

```
```mermaid
gantt
    title Throughput Linux arm64 (Total requests) 
    dateFormat  X
    axisFormat %s
    section Baseline
    This PR (4411) (9.618M)   : 0, 9617614
    master (9.820M)   : 0, 9819765
    benchmarks/2.33.0 (9.426M)   : 0, 9425603
    benchmarks/2.9.0 (9.842M)   : 0, 9842216

    section Automatic
    This PR (4411) (6.728M)   : 0, 6728466
    master (6.584M)   : 0, 6583758
    benchmarks/2.33.0 (6.803M)   : 0, 6802932

    section Trace stats
    master (6.837M)   : 0, 6837403
    benchmarks/2.33.0 (6.810M)   : 0, 6810447

    section Manual
    This PR (4411) (8.597M)   : 0, 8596619
    master (8.483M)   : 0, 8483262
    benchmarks/2.33.0 (8.435M)   : 0, 8434569

    section Manual + Automatic
    This PR (4411) (6.581M)   : 0, 6581038
    master (6.460M)   : 0, 6460201
    benchmarks/2.33.0 (6.465M)   : 0, 6465149

    section Version Conflict
    master (5.942M)   : 0, 5941596
    benchmarks/2.33.0 (5.807M)   : 0, 5806690

```
```mermaid
gantt
    title Throughput Windows x64 (Total requests) 
    dateFormat  X
    axisFormat %s
    section Baseline
    This PR (4411) (9.759M)   : 0, 9758680
    master (10.170M)   : 0, 10169653
    benchmarks/2.33.0 (9.426M)   : 0, 9426080
    benchmarks/2.9.0 (10.688M)   : 0, 10687842

    section Automatic
    This PR (4411) (6.990M)   : 0, 6990083
    master (7.216M)   : 0, 7216200
    benchmarks/2.33.0 (6.469M)   : 0, 6468939
    benchmarks/2.9.0 (7.829M)   : 0, 7828890

    section Trace stats
    master (7.287M)   : 0, 7287082
    benchmarks/2.33.0 (6.830M)   : 0, 6830314

    section Manual
    This PR (4411) (8.762M)   : 0, 8761857
    master (9.190M)   : 0, 9189578
    benchmarks/2.33.0 (8.108M)   : 0, 8108026

    section Manual + Automatic
    This PR (4411) (6.849M)   : 0, 6848899
    master (6.995M)   : 0, 6994675
    benchmarks/2.33.0 (6.856M)   : 0, 6856379

    section Version Conflict
    master (6.398M)   : 0, 6397504
    benchmarks/2.33.0 (5.893M)   : 0, 5893016

```
```mermaid
gantt
    title Throughput Linux x64 (ASM) (Total requests) 
    dateFormat  X
    axisFormat %s
    section Baseline
    This PR (4411) (7.565M)   : 0, 7565156
    master (7.563M)   : 0, 7563381
    benchmarks/2.33.0 (7.765M)   : 0, 7764607
    benchmarks/2.9.0 (7.769M)   : 0, 7768524

    section No attack
    This PR (4411) (2.122M)   : 0, 2121610
    master (2.137M)   : 0, 2137122
    benchmarks/2.33.0 (2.209M)   : 0, 2208725
    benchmarks/2.9.0 (3.174M)   : 0, 3174364

    section Attack
    This PR (4411) (1.821M)   : 0, 1820786
    master (1.861M)   : 0, 1860845
    benchmarks/2.33.0 (1.911M)   : 0, 1910909
    benchmarks/2.9.0 (2.518M)   : 0, 2517764

    section Blocking
    This PR (4411) (3.670M)   : 0, 3669642
    master (3.675M)   : 0, 3675283
    benchmarks/2.33.0 (3.861M)   : 0, 3860900

```

**Comment by NachoEchevarria** [comment]
Thank you for your feedback and reviews!

File: tracer/src/Datadog.Trace/Iast/HashBasedDeduplication.cs [hunk 2]
```diff
@@ -3,38 +3,53 @@
...
+        _cacheClearedTime = currentTime ?? DateTime.Now;
     }
 
     public static HashBasedDeduplication Instance { get; } = new();
 
-    public bool Add(Vulnerability vulnerability)
+    public bool Add(Vulnerability vulnerability, DateTime? addTime = null)
     {
         var hashCode = vulnerability.GetHashCode();
+        var currentTime = addTime ?? DateTime.Now;
```
Comment (andrewlock) [inline tracer/src/Datadog.Trace/Iast/HashBasedDeduplication.cs:35]: Same, I think this should use `DateTime.UtcNow`

File: tracer/src/Datadog.Trace/Iast/HashBasedDeduplication.cs [hunk 3]
```diff
@@ -3,38 +3,53 @@
...
 
     /// <summary>
     /// Initializes a new instance of the <see cref="HashBasedDeduplication"/> class.
     /// For testing only.
     /// Note that this API does NOT replace the global HashBasedDeduplication instance.
     /// </summary>
-    internal HashBasedDeduplication()
+    internal HashBasedDeduplication(DateTime? currentTime = null)
     {
+        _cacheClearedTime = currentTime ?? DateTime.Now;
```
[the remark this answers is not in the document]
Comment (NachoEchevarria) [reply tracer/src/Datadog.Trace/Iast/HashBasedDeduplication.cs:27]: Sorry, I did not notice this comment before merging. I guess that, potentially, a time correction for daylight savings could occur and affect the calculations. I will correct that. Thank you!

**Pull request merged**